### PR TITLE
Unify git archive (pt 2)

### PR DESCRIPTION
**Pull request by mollylogue**
*EDIT: This PR has been re-worked to just unify all git archive calls through the same function, `git.ArchiveReader`. If a sub-repo permissions checker is passed in and the repo has sub-repo perms enabled, an error will be returned. A follow-up PR will implement sub-repo filtering for this endpoint rather than returning an error*

- Previous behavior: return an error if `git.ArchiveReader` is called for a repo with sub-repo permissions
- New behavior: if sub-repo permissions are enabled and `git.ArchiveReader` is called, only include the files in the archive that the user has access to.
- Note: the places where `git.ArchiveReader` is called with a `nil` sub-repo permissions checker is just preserving the legacy behavior (these places didn't have sub-repo perms checking). If we want to add the checker in any of those locations I suggest we do it in a follow-up to this.
- Questions:  Edit:  ✅ means question was answered
-- Is it useful to still check if the repo has sub-repo permissions enabled before doing the `git ls-files` to avoid that extra check? We aren't worrying about that elsewhere (i.e. we only check if the sub repo permissions are enabled globally and not for the specific repo before trying to filter commits for a repo). ✅ 
-- Do we already have tooling around our zip archives for testing, etc? I just didn't want to write a bunch of new code to check the contents of the zip files in the unit tests if this already exists somewhere in the codebase. ✅ 
-- It was mentioned in the issue that the `git ls-files` command could happen on the gitserver side to limit the number of calls to gitserver. My concern with doing it this way is that we currently don't do any sub-repo permissions filtering in gitserver itself, leaving that to the `git` package which wraps the calls to gitserver. I don't know that we should diverge from that pattern just for this. ✅ 
-- What is the best way to test this at scale? Do we have an example repo with enough files that we could try this filtering on to see what performance is like? Could we do that in a production-like environment like dogfood? ✅ 

## Test plan
- Robust unit tests around different scenarios
- Manually set up sub-repo permissions for a test repo and request zip archive, then validate the archive only contains files the user has access to.
- Potentially writing an integration test for the above scenario?

### CODE REVIEW

**Comment by ryanslade** [comment]
> * Previous behavior: return an error if `git.ArchiveReader` is called for a repo with sub-repo permissions
> * New behavior: if sub-repo permissions are enabled and `git.ArchiveReader` is called, only include the files in the archive that the user has access to.
> * Note: the places where `git.ArchiveReader` is called with a `nil` sub-repo permissions checker is just preserving the legacy behavior (these places didn't have sub-repo perms checking). If we want to add the checker in any of those locations I suggest we do it in a follow-up to this.
> * Questions:
>   -- Is it useful to still check if the repo has sub-repo permissions enabled before doing the `git ls-files` to avoid that extra check? We aren't worrying about that elsewhere (i.e. we only check if the sub repo permissions are enabled globally and not for the specific repo before trying to filter commits for a repo).

I think it is worth doing in this case since in most cases we won't filter out any files and will then be sending a massive list of paths over to the archive command. 

>   -- Do we already have tooling around our zip archives for testing, etc? I just didn't want to write a bunch of new code to check the contents of the zip files in the unit tests if this already exists somewhere in the codebase.

We have some tests around the raw endpoint that may help:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a4c9837b04d7501a11c9e349aaf746205ce1530a/-/blob/cmd/frontend/internal/app/ui/raw_test.go?L126

>   -- It was mentioned in the issue that the `git ls-files` command could happen on the gitserver side to limit the number of calls to gitserver. My concern with doing it this way is that we currently don't do any sub-repo permissions filtering in gitserver itself, leaving that to the `git` package which wraps the calls to gitserver. I don't know that we should diverge from that pattern just for this.

I agree that for now we should take the performance hit of more round trips in order to keep our usage of sub-repo filtering consistent. 

>   -- What is the best way to test this at scale? Do we have an example repo with enough files that we could try this filtering on to see what performance is like? Could we do that in a production-like environment like dogfood?

You should be able to add some fake rules to any reasonably large repo, even locally. Something like our own `sourcegraph/sourcegraph` repo is already fairly large at about 8k files.

If you really want to stress test this you could try it with [linux](https://github.com/torvalds/linux) or [kubernetes](https://github.com/kubernetes/kubernetes)

**Comment by mollylogue** [comment]
I've run a test locally w/ sub-repo permissions filtering on for this repo and compared it against the usual behavior. I was getting numbers around 5 seconds for the call to the `raw` endpoint w/ filtering enabled vs 3.2-ish for the base case. I'm going to try on a bigger repo and see what happens/record timings.

**Comment by mollylogue** [comment]
On second thought, since that 431 is coming from gitserver, we could just add the ability to have a request body to gitserver and put the list of files in that rather than in a header.

**Comment by mollylogue** [comment]
So, I played around w/ moving the list of files to the request body, so the 431 http error is no longer happening. Now, however, we're running into `error: fork/exec /opt/homebrew/bin/git: argument list too long (stderr: "")` which is an issue with running the command itself. I'm investigating if there's a way to pass in a long argument list as a file (which is something that is doable for the `tar` command, for example). If there's no way to do that, I'm not sure we can handle filtering for repos of this size.

File: internal/vcs/git/archive.go [hunk 1]
```diff
@@ -21,27 +20,35 @@ const (
...
-		enabled, err := authz.SubRepoEnabledForRepoID(ctx, checker, repo.ID)
-		if err != nil {
+		if enabled, err := authz.SubRepoEnabledForRepo(ctx, checker, repo); err != nil {
 			return nil, errors.Wrap(err, "sub-repo permissions check:")
+		} else if enabled {
+			filteredFiles, err := LsFiles(ctx, checker, repo, api.CommitID(options.Treeish), options.Paths...)
+			if err != nil {
+				return nil, errors.Wrap(err, "LsFiles in ArchiveReader")
+			}
+			options.Paths = filteredFiles
```
Comment (keegancsmith) [inline internal/vcs/git/archive.go:40]: this list can be too large. I imagine sub-repo permissions are common on large monorepos, which means this will normally fail.

tar is a streaming format so you can filter out. This is more difficult for zip, but doable since in practice the format generated will look like a streaming file.

In theory you could batch teh calls, but that is also tricky for combining. Another approach is to implement a perm aware git archive on gitserver (go-git?). Another approach is always asking for tar which is streamable, but then creating a zip file here.

Taking a step back, where do we expose archives these days? My historical knowledge here is for extensions (which I think is no longer the case?) and src-cli's batch stuff? Maybe there is another approach here for those and we can hide the archive endpoint / fail it for sub-repo perms?

**Comment by keegancsmith** [comment]
FYI @chrismwendt has already done work around having long argument lists to git archive. See context in this PR https://github.com/sourcegraph/sourcegraph/pull/30658

File: internal/vcs/git/archive.go (hunk 1 above)
Comment (mollylogue) [reply internal/vcs/git/archive.go:40]: I like both of these ideas. I feel like requesting a tar from gitserver and then filtering/creating a zip file here makes the most sense. At the moment we aren't doing sub-repo permissions filtering inside gitserver, but rather in the `git` package that wraps calls to the gitserver api. The approach of requesting the tar and filtering here would be more in line with that approach, which is why I'm leaning in that direction.

Comment (mollylogue) [reply internal/vcs/git/archive.go:40]: I might for now (i.e. this PR) just fail early if the list of files is too long and then do a follow-up to implement the filtering in that case just for ease of review.

**Comment by mollylogue** [comment]
> I didn't know `POST` sends data via the body until Keegan told me:

I'm so confused as to why I was getting a 431 error before moving the list of files to the body on this line https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b05a272a521fe6bc03447cd6a22b135bfe3de182/-/blob/internal/gitserver/client.go?L331 🤔 

> The other limit is the length of the list of arguments to git, and for that I implemented batching in #30663

Is this just for tar archives? Either way, it would be super useful to move this functionality into a more central location like the `git` package or the gitserver client.

File: internal/vcs/git/archive.go (hunk 1 above)
Comment (mollylogue) [reply internal/vcs/git/archive.go:46]: I looked into go-git and it doesn't look like archiving is currently supported: https://sourcegraph.com/github.com/go-git/go-git/-/blob/COMPATIBILITY.md?L65  :/ 

Our original approach for sub-repo was to just fail anytime the archive endpoint is called when sub-repo perms are enabled. That might still be the best approach? I'm going back and forth on whether this is even worth it at this point. The reason I feel like it might be worth it is there are other areas of the codebase where we're attempting to call `git archive` with a long list of files (I think just symbols), so we're already needing to figure out how to handle this issue.

**Comment by chrismwendt** [comment]
The arg limit is enforced by the OS, so it applies to all commands including git.

> Chris, how would you feel about that code moving into the git package, and letting it be handled there? It would simplify the code in symbols and then any caller of git.ArchiveReader would get the benefit of being able to request a large list of paths (for a tar archive request, will have to handle zip differently).

Sounds good to me, the symbols service doesn't really need to know or care how the files are fetched. It doesn't even require the tar format - all it cares about is the file contents, so there's leeway in the implementation to use `git archive` or `git cat-file --batch` or something else. `cat-file` could be useful because you can stream filenames in via stdin and avoid the arg limit.

File: internal/vcs/git/archive.go (hunk 1 above)
Comment (keegancsmith) [reply internal/vcs/git/archive.go:46]: Yeah internally I believe everywhere we call archive it isn't on behalf of a user, but to analyse the repo to create an index / stats / etc. Those services then need to respect sub repo permissions themselves (And already do?). For example symbols does this as you point out, but will do it as "root". It will then filter out the symbols a user can't see for a user request.

I'm not actually sure where archives are used externally, the only memory I have is src-cli using them for batch changes. cc @eseliger who can comment on if archive is still needed. If not, I imagine just failing if the request is for a user makes a lot of sense to me since it is much simpler than actually making this work.

> I looked into go-git and it doesn't look like archiving is currently supported

Yeah you would have to implement this yourself, which would be relatively straightforward using its tree walker stuff. But to be honest, it would be easier to just implement the idea of always using tar and synthesizing a zip if requested.

**Comment by mollylogue** [comment]
Here's where I'm at with this:

The current usage of `git archive` is largely limited to internal actors or services that have sub-repo filtering done at a higher level. I believe the one exception to this is the `raw` endpoint. For this iteration, I think it's probably sufficient to just disable the usage of that endpoint in the case of sub-repo permissions being enabled (or alternatively allow usage if the list of files is small enough?), and then in the other cases just pass a `nil` sub-repo perms checker and add a comment that the filtering is done at a higher level. That's essentially what I already have in this PR (aside from the comments and potentially just disabling the `raw` endpoint entirely for the sub-repo perms case)

I _do_ think it's worthwhile to craft a more robust solution for archiving (in the case of sub-repo permissions filtering, or in the case where we need to request a long list of files). I think it makes sense to do that separately. I'm going to investigate several options on this front and probably do a quick write-up on each before diving into implementing one. 

TLDR: I'd like to get reviews on this as is, with the assumption that the `raw` endpoint is the only location that will currently be using this filtering code and it will fail for a list of files that is too long.

File: cmd/searcher/main.go [hunk 2]
```diff
@@ -71,7 +71,7 @@ func main() {
 	service := &search.Service{
 		Store: &store.Store{
 			FetchTar: func(ctx context.Context, repo api.RepoName, commit api.CommitID) (io.ReadCloser, error) {
-				return git.ArchiveReader(ctx, repo, gitserver.ArchiveOptions{Treeish: string(commit), Format: git.ArchiveFormatTar})
+				return git.ArchiveReader(ctx, nil, repo, gitserver.ArchiveOptions{Treeish: string(commit), Format: git.ArchiveFormatTar})
```
[the remark this answers is not in the document]
Comment (keegancsmith) [reply cmd/searcher/main.go:74]: yes. Filtering happens in a higher layer. I can't remember exactly where this ended up, but likely is done inside of sourcegraph-frontend. I think @ryanslade implemented that for search.

**Comment by keegancsmith** [comment]
> I believe the one exception to this is the `raw` endpoint. For this iteration, I think it's probably sufficient to just disable the usage of that endpoint in the case of sub-repo permissions being enabled (or alternatively allow usage if the list of files is small enough?),

That sounds good to me. I'd just disable. Failing depending on file size is more complicated to implement and harder to understand/debug when it happens.

> and then in the other cases just pass a `nil` sub-repo perms checker and add a comment that the filtering is done at a higher level. That's essentially what I already have in this PR (aside from the comments and potentially just disabling the `raw` endpoint entirely for the sub-repo perms case)

For the background cases, if I am not mistaken there is no associated user. Maybe if there is no user that is when you use this behaviour? If there is a user associated with the request, you fail. I'm not sure if that is safe enough. I'd confirm by ensuring there are no intermediate caching with the raw endpoint. In the past it relied on a local disk cache, but I think we removed that.

> I _do_ think it's worthwhile to craft a more robust solution for archiving (in the case of sub-repo permissions filtering, or in the case where we need to request a long list of files). I think it makes sense to do that separately. I'm going to investigate several options on this front and probably do a quick write-up on each before diving into implementing one.

Agreed separately makes sense.

> TLDR: I'd like to get reviews on this as is, with the assumption that the `raw` endpoint is the only location that will currently be using this filtering code and it will fail for a list of files that is too long.

As mentioned above, I think the safer implementation is to just always fail. If you'd rather stick with the list size implementation, happy to review this approach instead.

File: internal/database/sub_repo_perms_store.go [hunk 3]
```diff
@@ -204,3 +205,29 @@ AND external_service_type IN (%s)
 	}
 	return *exists, nil
 }
+
+// RepoSupported returns true if repo has sub-repo permissions
+// (i.e. it is private and its type is one of the SubRepoSupportedCodeHostTypes)
+func (s *subRepoPermsStore) RepoSupported(ctx context.Context, repo api.RepoName) (bool, error) {
+	q := sqlf.Sprintf(`
+SELECT EXISTS(
```
Comment (tsenart) [inline internal/database/sub_repo_perms_store.go:213]: Do we need the outer select?

Comment (mollylogue) [reply internal/database/sub_repo_perms_store.go:213]: I was honestly copying what we do for doing this check w/ repo id: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/sub_repo_perms_store.go?L184
I'm guessing the reason for the outer select is to that the query returns a bool which this function then returns? You're probably right that it's not necessary (we could just check if anything is returned from the first select and return true if so)

Comment (ryanslade) [reply internal/database/sub_repo_perms_store.go:213]: Isn't this the canonical way to check for the existence of any rows that match a query?

Also, you could use `basestore.ScanFirstBool` to reduce this code a bit

Comment (ryanslade) [reply internal/database/sub_repo_perms_store.go:213]: I decided to make the change in the other place as an example:
https://github.com/sourcegraph/sourcegraph/pull/32934

**Review by ryanslade — APPROVED** [review]

File: internal/database/sub_repo_perms_store.go (hunk 3 above)
Comment (tsenart) [reply internal/database/sub_repo_perms_store.go:213]: > Isn't this the canonical way to check for the existence of any rows that match a query?

Not that I was aware of! But it works. I'd personally find it simpler to avoid the sub-query with something like:

```sql
SELECT COUNT(*)
FROM repo
WHERE name = %s
AND private = TRUE
AND external_service_type IN (%s)
LIMIT 1;
```

And then compare that count to zero to test if it exists. But this is just ergonomics and personal taste, not a big deal at all.
